### PR TITLE
Simplify: Remove legacy login password

### PR DESCRIPTION
**Pull request by DanVanAtta**
## Overview
We have code to support authenticating against either old or new
password columns (old is md5, new is bcrypt). This update drops the old
'password' column code and we use just the new bcrypt password column.

## Functional Changes
- none, we should be migrated to bcrypt entirely by now.  

## Manual Testing Performed
- using a local lobby + DB:
  - created a new account with current master, restarted lobby with this branch and verified could log in
  - created new account with this branch, verified could log in
  - created account with this branch, updated password, verified could log in with the updated password 

Checking on prod DB, we always have a bcrypt_password where we have a password:
```
# select count(*) from ta_users where password is null and bcrypt_password is null
 count 
-------
     0

# select count(*) from ta_users 
 count 
-------
11369
```

## Additional Review Notes
- removed 'mask' function on HashPassword, it was used for logging, we can not log the value at all and thereby not need to mask it.

### CODE REVIEW

**Comment by ssoloff** [comment]
> Checking on prod DB, we always have a bcrypt_password where we have a password:
> ```
> # select count(*) from ta_users where password is null and bcrypt_password is null
>  count 
> -------
>      0
> ```

I believe the query to check if a user has both a legacy password and a bcrypted password is 

```
ta_users=# select count(*) from ta_users where password is not null and bcrypt_password is not null;
 count 
-------
  9387
(1 row)
```

If you're checking for the number of users who have NOT migrated to bcrypt, that would be

```
ta_users=# select count(*) from ta_users where password is not null and bcrypt_password is null;
 count 
-------
  1211
(1 row)
```

**Comment by RoiEXLab** [comment]
Good catch @ssoloff 
I should've read the whole query (to my excuse I read the description on my phone, so it was half cut off 😅).
But we can probably call this project quit, so I'm gonna close this PR.

In case you want to revive this project by simply removing all users that haven't migrated yet, assuming they are inactive for years now, feel free to reopen or create a new PR.

**Comment by DanVanAtta** [comment]
> But we can probably call this project quit, so I'm gonna close this PR.

Are you referring to the migration to bcrypt @RoiEXLab ?

The context of how I got here was after a number of refactorings that started to blow up in scope. The login code has a lot of calls into DB layer, it's particularly complex and IMO with the promise of https there would be a plan to remove it.

@ssoloff good catch on the query, I got to this from the code-perspective. It looks like we still create and use md5 passwords without a bcrypt. I see a login as recent as yesterday that contained no bcrypt password, some accounts were created recently, others long off.

I think that means removing the 'bcrypt' side of this would have been the better move :confused:

**Comment by DanVanAtta** [comment]
I do want this to land as this was a pretty difficult effort to get these 400 lines of code removed. Assigning to the three of us to discuss.

**Comment by RoiEXLab** [comment]
@DanVanAtta
The current code is already only in place for backwards compatibility, so I was assuming that you were creating this PR because you noticed that it was no longer needed. Because it is still "needed" in the sense that there are still accounts without a bcrypt hashed. That's why I closed the PR.

Unless you simply want to drop all of the users that don't support bcrypt yet I don't see how this PR will achieve your goal.

By the time of writing this I'm not sure what exactly is your goal here. I'm almost certain that there's a misunderstanding at some point, so I'm sure that your motivation for this change will help me understand 👍

File: lobby/src/main/java/org/triplea/lobby/server/db/UserController.java [hunk 1]
```diff
@@ -68,33 +59,18 @@ public void updateUser(final DBUser user, final HashedPassword hashedPassword) {
...
-          ps2.setString(1, user.getName());
-          ps2.execute();
-        }
-      }
       con.commit();
     } catch (final SQLException e) {
-      throw newDatabaseException(String.format("Error updating name: %s, email: %s, (masked) pwd: %s",
-          user.getName(), user.getEmail(), hashedPassword.mask()), e);
+      throw newDatabaseException(String.format("Error updating name: %s, email: %s",
+          user.getName(), user.getEmail()), e);
```
[the remark this answers is not in the document]
Comment (DanVanAtta) [reply lobby/src/main/java/org/triplea/lobby/server/db/UserController.java]: A few reasons actually:
- code attached to methods to support only logging is a bit of a code smell
- people reading the logs will not be confused by not seeing a lack of password. If we think the information *would* be very useful to know about errors, then we can handle it at the time of logging. Given we are dealing with a password, not even operating on it keeps the local scope smaller and easier to audit. It's already very questionable that we log email and full username (it can be recommend not logging full username for the cases when a user had the wrong fields focused and types their password as their username).

**Comment by DanVanAtta** [comment]
This is unexpected about the bcrypt migration:

```
select max(lastlogin)
from ta_users where password is not null and bcrypt_password is null;

            max             
----------------------------
 2019-02-22 14:05:37.989-05
```
I was expecting any login or new account to create a bcrypt_password.

I can see a lot of argument for users rotating passwords that were ever stored in md5crypt. Given the migration we tried is still at square 1 :face_with_head_bandage: , unless there are other really good proposals, I think this means we'll need to have users recreate their lobby accounts once again on the next release.

**Comment by DanVanAtta** [comment]
> I'm almost certain that there's a misunderstanding at some point, so I'm sure that your motivation for this change will help me understand

Two goals:
1) Reduce core complexity: https://github.com/triplea-game/triplea/pull/4703/files#diff-ae2d9b2eb6c5f8e068cf0ff3858014d9L238
2) Reduce calls and dependency on DB layer. There is a sheer volume of calls that go to the backend here, any work to move/update that backend code winds up updating a ton of code to support legacy.

**Comment by DanVanAtta** [comment]
There is a third goal, that is to eventually migrate the login code to a new DB schema, behind http server, and to have the password not be de-cryptable but instead use a standard approach where we get a hased value on lobby side, we hash that and see if it exists in DB.

If we're going to replace user accounts on next release, then this PR I think is viable. Otherwise, we need a plan to be proposed (in either case the login code would all be replaced, but if the right simplification is to first instead remove bcrypt - then this update probably should be redone.)

**Comment by DanVanAtta** [comment]
> Unless you simply want to drop all of the users that don't support bcrypt

This might be the disconnect. Why would this be the case? Is our last latest release that old? Or is it that we have not forced users to upgrade?

I'm guessing it's the very latter, which would be good news.

**Comment by DanVanAtta** [comment]
TL;DR
I think we are actually good to go here, but we do need to get hopping on forcing a min version upgrade to the last release.

@RoiEXLab, agree?

**Comment by DanVanAtta** [comment]
@ssoloff re: DB stat, it was a bit late when I "threw" that in, a good time when sleep is better compared to getting a PR out the door. On the other hand the better understanding has been useful.

Regardless, the DB data about migration is inconsequential to this update, that is fixed by us getting on the ball and forcing an engine update to the latest release.

**Comment by RoiEXLab** [comment]
@DanVanAtta If we could force people to upgrade to the latest stable, our problems would be resolved, correct.
I think the disconnect was that I was assuming the PR should wait until the migration finished, but I wasn't considering that we could just wait with the deployment.
However in the latter case we'd have to implement some sort of mechanism to "force the upgrade" on 1.9.
Not sure how to do that.

**Comment by DanVanAtta** [comment]
Okay, it goes back to the 5 digit version number hackery, I see the problem now.

Perhaps we could do this, create a special 'v1.9.0.1' version from the last release and mark it as latest. The increment on patch should AFAIK keep save and network games working. We'll want to investigate more if this idea seems viable. Then from there we can toggle the min-engine version to be v1.9.0.1. 

Getting people synced up on the latest release is a really good thing. I'm thinking this is a good approach rather than waiting for this to play out over multiple releases. Given that we are not fast about migrations, if we continue waiting with such long project increments then I really fear for what we can accomplish.

On the side of this is still a question if we should go ahead and force everyone to update their password anyways. Strictly speaking that is the most secure thing to do even if we have no reason to think our DB data has been compromised. If we decide to go that route, then a seamless password transition could be moot point.

**Comment by RoiEXLab** [comment]
@DanVanAtta The most elegant solution to this problem I have seen so far is how the A&A Forum was migrated to NodeBB.
In order to migrate the password, everyone had to reset their password before logging in after the migration.
Of course this would require a password reset function first, which we want anyways.

**Comment by DanVanAtta** [comment]
Thanks @RoiEXLab , I think that means we keep both password columns, then prompt to recreate password if an old authentication is used.
